crates/sui-framework-build/src/compiled_package.rs

```diff
@@ -0,0 +1,273 @@
+    /// for publishing or static analysis
+    pub fn get_modules(&self) -> Vec<&CompiledModule> {
+        self.package
+            .root_modules()
+            .map(|m| match &m.unit {
+                CompiledUnitEnum::Module(m) => &m.module,
+                CompiledUnitEnum::Script(_) => unimplemented!("Scripts not supported in Sui Move"),
+            })
+            .collect()
+    }
+
+    /// Return all of the bytecode modules that this package depends on (both directly and transitively)
+    /// Note: these are not topolically sorted by dependency-
+    pub fn get_dependent_modules(&self) -> Vec<&CompiledModule> {
+        self.package
+            .deps_compiled_units
+            .iter()
+            .map(|(_, m)| match &m.unit {
+                CompiledUnitEnum::Module(m) => &m.module,
+                CompiledUnitEnum::Script(_) => unimplemented!("Scripts not supported in Sui Move"),
+            })
+            .collect()
+    }
+
+    /// Return the bytecode modules in this package, topologically sorted in dependency order
+    /// This is the function to call if you would like to publish or statically analyze the modules
+    pub fn get_dependency_sorted_modules(&self) -> SuiResult<Vec<CompiledModule>> {
+        let compiled_modules = self.package.root_modules_map();
+        // Collect all module IDs from the current package to be
+        // published (module names are not sufficient as we may
+        // have modules with the same names in user code and in
+        // Sui framework which would result in the latter being
+        // pulled into a set of modules to be published).
+        // For each transitive dependent module, if they are not to be published,
+        // they must have a non-zero address (meaning they are already published on-chain).
+        let self_modules: HashSet<ModuleId> = compiled_modules
+            .iter_modules()
+            .iter()
+            .map(|m| m.self_id())
+            .collect();
+        if let Some(m) = self
+            .package
+            .deps_compiled_units
+            .iter()
+            .find_map(|(_, unit)| match &unit.unit {
+                CompiledUnit::Module(NamedCompiledModule { module: m, .. })
+                    if !self_modules.contains(&m.self_id())
+                        && m.self_id().address() == &AccountAddress::ZERO =>
+                {
+                    Some(m)
+                }
+                _ => None,
+            })
+        {
+            return Err(SuiError::ModulePublishFailure { error: format!("Dependent modules must have been published on-chain with non-0 addresses, unlike module {:?}", m.self_id()) });
+        }
+        Ok(self
+            .package
+            .all_modules_map()
+            .compute_dependency_graph()
+            .compute_topological_order()
+            .unwrap()
+            .filter(|m| self_modules.contains(&m.self_id()))
+            .cloned()
```

Comment:
Do we need to always clone here?

Reply:
Unfortunately yes--`compute_topological_order` returns a vector of references into a vector that is not ordered. I don't think it's possible to return an owned value without either cloning or shuffling the vector to make it respect the ordering.